File: 3635-earliest-finish-time-for-land-and-water-rides-ii/3635-earliest-finish-time-for-land-and-water-rides-ii.go
func earliestFinishTime(landStartTime []int, landDuration []int, waterStartTime []int, waterDuration []int) int {
	// 1. 땅 -> 물 순서로 탈 때의 최솟값 계산
	minLandEnd := math.MaxInt32
	for i := 0; i < len(landStartTime); i++ {
		end := landStartTime[i] + landDuration[i]
		if end < minLandEnd {
			minLandEnd = end
		}
	}

	ans1 := math.MaxInt32
	for j := 0; j < len(waterStartTime); j++ {
		// 땅이 끝난 시간과 물 놀이기구 시작 시간 중 더 늦은 시간에 시작
		start := minLandEnd
		if waterStartTime[j] > start {
			start = waterStartTime[j]
		}
		end := start + waterDuration[j]
		if end < ans1 {
			ans1 = end
		}
	}

	// 2. 물 -> 땅 순서로 탈 때의 최솟값 계산
	minWaterEnd := math.MaxInt32
	for j := 0; j < len(waterStartTime); j++ {
		end := waterStartTime[j] + waterDuration[j]
		if end < minWaterEnd {
			minWaterEnd = end
		}
	}

	ans2 := math.MaxInt32
	for i := 0; i < len(landStartTime); i++ {
		// 물이 끝난 시간과 땅 놀이기구 시작 시간 중 더 늦은 시간에 시작
		start := minWaterEnd
		if landStartTime[i] > start {
			start = landStartTime[i]
		}
		end := start + landDuration[i]
		if end < ans2 {
			ans2 = end
		}
	}

	// 두 경우 중 가장 이른 종료 시간 반환
	if ans1 < ans2 {
		return ans1
	}
	return ans2
}

File: 3635-earliest-finish-time-for-land-and-water-rides-ii/3635-earliest-finish-time-for-land-and-water-rides-ii.java
class Solution {
    public int earliestFinishTime(int[] landStartTime, int[] landDuration, int[] waterStartTime, int[] waterDuration) {
        long plan1 = findEarliestFinish(landStartTime, landDuration, waterStartTime, waterDuration);
        long plan2 = findEarliestFinish(waterStartTime, waterDuration, landStartTime, landDuration);
        return (int) Math.min(plan1, plan2); // Cast the final minimum long value to int
    }

    private long findEarliestFinish(int[] firstStart, int[] firstDuration, int[] secondStart, int[] secondDuration) {
        long minFinish = Long.MAX_VALUE;
        
        for (int i = 0; i < firstStart.length; i++) {
            long finishTime = (long) firstStart[i] + firstDuration[i];
            minFinish = Math.min(minFinish, finishTime);
        }

        long earliestFinishTime = Long.MAX_VALUE;
        for (int i = 0; i < secondStart.length; i++) {
            long startSecond = Math.max((long) secondStart[i], minFinish);
            long finishSecond = startSecond + secondDuration[i];
            earliestFinishTime = Math.min(earliestFinishTime, finishSecond);
        }

        return earliestFinishTime;
    }
}
